Packages/ClientRuntime/Sources/Waiters/WaiterUtils.swift

```diff
@@ -0,0 +1,120 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+public extension Array {
+
+    /// Flattens an array of arrays down to single depth.
+    /// - Parameter transform: A block that transforms the element(s) of this array into an array containing type T.
+    /// - Returns: A single-depth array of type T.
+    func flattenIfPossible<T>(_ transform: (Element) throws -> Array<T>) rethrows -> Array<T> {
+        return try flatMap { try transform($0) }
+    }
+
+    /// Returns the receiver mapped to an array of a different, non-array type using the provided transform block.
+    /// - Parameters:
+    ///   - unused: An unused integer parameter which is used to affect the precedence of selection of overloads by the Swift compiler.
+    ///   See discussion in the implementation.
+    ///   - transform: A block that transforms each element of the array into a type that is not itself an Array.
+    /// - Returns: The receiver, mapped by the `transform` block.
+    func flattenIfPossible<T>(_ unused: Int = 0, _ transform: (Element) throws -> T) rethrows -> Array<T> {
+        // Adding an unused argument with a default value to this function's signature allows it to still be
+        // an overload of the flattenIfPossible() implementation above, but ranks it lower when
+        // the compiler selects an overload.
+        // Without the unused argument, the compiler uses this implementation instead of the one above when
+        // the element of an array is also an array, which is not desired behavior.
+        // See: https://forums.swift.org/t/how-to-specify-which-is-the-default-function-when-functions-use-the-same-name-like-array-reversed/39168/5
+        // In the future, if Swift allows us to manually rank overloads or otherwise fixes this issue,
+        // we should eliminate this param & use the new mechanism for overload resolution instead.
+        return try map { try transform($0) }
+    }
+}
+
+/// A value type that is used to facilitate comparison of values returned by JMESPath types, in accordance with the JMESPath spec.
+public enum JMESValue: Equatable, Comparable {
```

Comment:
Can we place this type in it's own file?

Reply:
This type is great at it's job, but it is an implementation detail. Given that, it'd be great to keep this internal and not expose it publicly, especially since JMES is a real thing and we are only defining this one small piece of it. 

I'd love if we exposed some function instead that used this type under the hood. Something like 
```swift
extension WaiterConfiguration.Acceptor {

    public enum Comparator {
        case equals
        case lessThan
    }
    public static func compareValues(_ lhs: Any?, _ rhs: Any?, comparator: Comparator) -> Bool {
            let lhsValue = JMESValue(lhs)
            let rhsValue = JMESValue(rhs)
            switch comparator {
            case .equals: 
                return lhsValue == rhsValue
            case .lessThan:
                return lhsValue < rhsValue
            }
    }
}
```

And then you just need to add an initializer for JMESValue that takes an `Any?`. Could even add an enum case for "unsupported type" and that way you can define equals and less than explicitly for those cases.